keps/sig-api-machinery/20190228-priority-and-fairness.md

```diff
@@ -233,7 +233,589 @@ yet but we think may be interesting to consider in the future.
+```yaml
+kind: RequestPriority
+meta:
+  name: workload-high
+spec:
+  assuredConcurrencyShares: 30
+  queues: 128
+  handSize: 6
+  queueLengthLimit: 100
+```
+
+For requests from controllers processing workload.
+```yaml
+kind: RequestPriority
+meta:
+  name: workload-low
+spec:
+  catchAll: true
+  assuredConcurrencyShares: 100
+  queues: 128
+  handSize: 6
+  queueLengthLimit: 100
+```
+
+Some flow schemata.
+
+```yaml
+kind: FlowSchema
+meta:
+  name: system-top
+spec:
+  requestPriority:
+    name: system-top
+  match:
+  - and: # writes by admins (does this cover loopback too?)
+    - superSet:
+      field: groups
+      set: [ "system:masters" ]
+```
+
+```yaml
+kind: FlowSchema
+meta:
+  name: system-high
+spec:
+  requestPriority:
+    name: system-high
+  flowDistinguisher:
+    source: user
+    # no transformation in this case
+  match:
+  - and: # heartbeats by nodes
+    - superSet:
+      field: groups
+      set: [ "system:nodes" ]
+    - equals:
+      field: resource
+      value: nodes
+  - and: # kubelet and kube-proxy ops on system objects
+    - superSet:
+      field: groups
+      set: [ "system:nodes" ]
+    - equals:
+      field: namespace
+      value: kube-system
+  - and: # leader elections for system controllers
+    - patternMatch:
+      field: user
+      pattern: system:controller:.*
+    - inSet:
+      field: resource
+      set: [ "endpoints", "configmaps", "leases" ]
+    - equals:
+      field: namespace
+      value: kube-system
+```
+
+```yaml
+kind: FlowSchema
+meta:
+  name: system-low
+spec:
+  matchingPriority: 900
+  requestPriority:
+    name: system-low
+  flowDistinguisher:
+    source: user
+    # no transformation in this case
+  match:
+  - and: # the garbage collector
+    - equals:
+      field: user
+      value: system:controller:garbage-collector
+```
+
+```yaml
+kind: FlowSchema
+meta:
+  name: workload-high
+spec:
+  requestPriority:
+    name: workload-high
+  flowDistinguisher:
+    source: namespace
+    # no transformation in this case
+  match:
+  - and: # users using kubectl
+    - notPatternMatch:
+      field: user
+      pattern: system:serviceaccount:.*
+```
+
+```yaml
+kind: FlowSchema
+meta:
+  name: workload-low
+spec:
+  matchingPriority: 9999
+  requestPriority:
+    name: workload-high
+  flowDistinguisher:
+    source: namespace
+    # no transformation in this case
+  match:
+  - and: [ ] # match everything
+```
+  
+Following is a FlowSchema that might be used for the requests by the
+aggregated apiservers of
+https://github.com/MikeSpreitzer/kube-examples/tree/add-kos/staging/kos
+to create TokenReview and SubjectAccessReview objects.
+
+
+```
+kind: FlowSchema
+meta:
+  name: system-top
+spec:
+  matchingPriority: 900
+  requestPriority:
+    name: system-top
+  flowDistinguisher:
+    source: user
+    # no transformation in this case
+  match:
+  - and:
+    - inSet:
+      field: resource
+      set: [ "tokenreviews", "subjectaccessreviews" ]
+    - superSet:
```

Comment:
My understanding of all the possible 'Match' directives is here: https://gist.github.com/aaron-prindle/b53d3d3687ef442fe113325db5fd4dff

Is my understanding of SuperSet correct in that the value needs to be in the set of the values in the field?  I'm not familiar with how Groups are represented in k8s, would this be analagous to the value being a substring of the field?

Reply:
Regarding superset: note that authentication produces a _set_ of strings, each of which names one of the groups that the user is authenticated to belong to.  Confusingly, this _set_ is often called simply `group`.

The general idea of the superset match is that when the authenticated request attribute identified by `field` is a set --- and let's call it set S1 --- and the match specified anothet set --- and let's call this one set S2 --- then we have a match if and only if every member of S2 is also a member of S1.  This is the reverse of subset, which would be true if and only if every member of S1 is also a member of S2.

So, for example, if the set of authenticated group strings is {"g1", "g2", "g4"} and the match specifies `field: groups, set: ["g1", "g2"]` then this is a successful match because both g1 and g2 are in the authenticated set.  OTOH, if the match said `field: groups, set: ["g2", "g3"]` then we would NOT have a match.

Reply:
ah makes sense, thanks you